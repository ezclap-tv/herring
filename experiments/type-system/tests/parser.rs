use compiler::*;

macro_rules! assert_ok {
  ($name:ident $source:literal) => {
    #[test]
    fn $name() {
      let source = $source;
      let name = stringify!($name);
      match parser::parse(&source) {
        Ok(ast) => {
          insta::assert_yaml_snapshot!(ast);
        }
        Err(errors) => {
          let mut buf = String::new();
          error::report(&source, &errors, &mut buf);
          panic!("{name} should have parsed sucessfully, but it failed:\n{buf}");
        }
      }
    }
  };
}

macro_rules! assert_error {
  ($name:ident $source:literal) => {
    #[test]
    fn $name() {
      let source = $source;
      let name = stringify!($name);
      match parser::parse(&source) {
        Ok(_) => {
          panic!("{name} should have failed to parse, but it was successful");
        }
        Err(errors) => {
          insta::assert_yaml_snapshot!(errors);
        }
      }
    }
  };
}

assert_ok!(full
r#"
// variables
let v = 10;

let v: int = 10; // int
let v: bool = true; /// bool
let v: str = "test"; /// str
let v: {x:int, y:int} = {x: 10, y: 10}; /// record

// functions
// single argument
// parameter and return type annotations are required
let fib n: int -> int =
  if n < 2 then n
  else n * fib (n - 1)
  ;

// function calls
fib (v.x);

let print_square n: int -> int =
  let r = intToStr (n * n) in
  print (n + " * " + n + " = " + r)
  ;

let unwrap r: {v:int} -> int = r.v;

print (unwrap {v:10});
"#);

assert_ok!(simple_expr
r#"
v || v;
v && v;
v == v;
v != v;
v < v;
v > v;
v <= v;
v >= v;
v + v;
v - v;
v * v;
v / v;
v % v;
v ** v;
~v;
~~v;
~~~~~~~~~v;
~~~~~~~~~~v;
v.f;
f (v);
f {v:v};
v.f (v);
v.f {v:v};
10;
true;
false;
"test";
{v:v};
v;
(((v)));
"#);

assert_error!(top_level_without_semi
r#"
let x = 10
let y = 5
"#
);

assert_ok!(types
r#"
let v: T = v;
let v: {v: T} = v;
let v: {v: T, v: T} = v;
let v: (((T))) = v;
"#);
