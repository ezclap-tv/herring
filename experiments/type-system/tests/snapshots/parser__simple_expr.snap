---
source: tests/parser.rs
expression: ast
---
- Binary:
    op: Or
    lhs:
      Use:
        ident:
          lexeme: v
          span:
            start: 1
            end: 2
          kind: Ident
    rhs:
      Use:
        ident:
          lexeme: v
          span:
            start: 6
            end: 7
          kind: Ident
- Binary:
    op: And
    lhs:
      Use:
        ident:
          lexeme: v
          span:
            start: 9
            end: 10
          kind: Ident
    rhs:
      Use:
        ident:
          lexeme: v
          span:
            start: 14
            end: 15
          kind: Ident
- Binary:
    op: Equal
    lhs:
      Use:
        ident:
          lexeme: v
          span:
            start: 17
            end: 18
          kind: Ident
    rhs:
      Use:
        ident:
          lexeme: v
          span:
            start: 22
            end: 23
          kind: Ident
- Binary:
    op: NotEqual
    lhs:
      Use:
        ident:
          lexeme: v
          span:
            start: 25
            end: 26
          kind: Ident
    rhs:
      Use:
        ident:
          lexeme: v
          span:
            start: 30
            end: 31
          kind: Ident
- Binary:
    op: LessThan
    lhs:
      Use:
        ident:
          lexeme: v
          span:
            start: 33
            end: 34
          kind: Ident
    rhs:
      Use:
        ident:
          lexeme: v
          span:
            start: 37
            end: 38
          kind: Ident
- Binary:
    op: GreaterThan
    lhs:
      Use:
        ident:
          lexeme: v
          span:
            start: 40
            end: 41
          kind: Ident
    rhs:
      Use:
        ident:
          lexeme: v
          span:
            start: 44
            end: 45
          kind: Ident
- Binary:
    op: LessEqual
    lhs:
      Use:
        ident:
          lexeme: v
          span:
            start: 47
            end: 48
          kind: Ident
    rhs:
      Use:
        ident:
          lexeme: v
          span:
            start: 52
            end: 53
          kind: Ident
- Binary:
    op: GreaterEqual
    lhs:
      Use:
        ident:
          lexeme: v
          span:
            start: 55
            end: 56
          kind: Ident
    rhs:
      Use:
        ident:
          lexeme: v
          span:
            start: 60
            end: 61
          kind: Ident
- Binary:
    op: Add
    lhs:
      Use:
        ident:
          lexeme: v
          span:
            start: 63
            end: 64
          kind: Ident
    rhs:
      Use:
        ident:
          lexeme: v
          span:
            start: 67
            end: 68
          kind: Ident
- Binary:
    op: Sub
    lhs:
      Use:
        ident:
          lexeme: v
          span:
            start: 70
            end: 71
          kind: Ident
    rhs:
      Use:
        ident:
          lexeme: v
          span:
            start: 74
            end: 75
          kind: Ident
- Binary:
    op: Mult
    lhs:
      Use:
        ident:
          lexeme: v
          span:
            start: 77
            end: 78
          kind: Ident
    rhs:
      Use:
        ident:
          lexeme: v
          span:
            start: 81
            end: 82
          kind: Ident
- Binary:
    op: Div
    lhs:
      Use:
        ident:
          lexeme: v
          span:
            start: 84
            end: 85
          kind: Ident
    rhs:
      Use:
        ident:
          lexeme: v
          span:
            start: 88
            end: 89
          kind: Ident
- Binary:
    op: Rem
    lhs:
      Use:
        ident:
          lexeme: v
          span:
            start: 91
            end: 92
          kind: Ident
    rhs:
      Use:
        ident:
          lexeme: v
          span:
            start: 95
            end: 96
          kind: Ident
- Binary:
    op: Power
    lhs:
      Use:
        ident:
          lexeme: v
          span:
            start: 98
            end: 99
          kind: Ident
    rhs:
      Use:
        ident:
          lexeme: v
          span:
            start: 103
            end: 104
          kind: Ident
- Unary:
    op: Negate
    rhs:
      Use:
        ident:
          lexeme: v
          span:
            start: 107
            end: 108
          kind: Ident
- Use:
    ident:
      lexeme: v
      span:
        start: 112
        end: 113
      kind: Ident
- Unary:
    op: Negate
    rhs:
      Use:
        ident:
          lexeme: v
          span:
            start: 124
            end: 125
          kind: Ident
- Use:
    ident:
      lexeme: v
      span:
        start: 137
        end: 138
      kind: Ident
- Access:
    target:
      Use:
        ident:
          lexeme: v
          span:
            start: 140
            end: 141
          kind: Ident
    field:
      lexeme: f
      span:
        start: 142
        end: 143
      kind: Ident
- Call:
    func:
      Use:
        ident:
          lexeme: f
          span:
            start: 145
            end: 146
          kind: Ident
    arg:
      Use:
        ident:
          lexeme: v
          span:
            start: 148
            end: 149
          kind: Ident
- Call:
    func:
      Use:
        ident:
          lexeme: f
          span:
            start: 152
            end: 153
          kind: Ident
    arg:
      Lit:
        Record:
          - ident:
              lexeme: v
              span:
                start: 155
                end: 156
              kind: Ident
            value:
              Use:
                ident:
                  lexeme: v
                  span:
                    start: 157
                    end: 158
                  kind: Ident
- Call:
    func:
      Access:
        target:
          Use:
            ident:
              lexeme: v
              span:
                start: 161
                end: 162
              kind: Ident
        field:
          lexeme: f
          span:
            start: 163
            end: 164
          kind: Ident
    arg:
      Use:
        ident:
          lexeme: v
          span:
            start: 166
            end: 167
          kind: Ident
- Call:
    func:
      Access:
        target:
          Use:
            ident:
              lexeme: v
              span:
                start: 170
                end: 171
              kind: Ident
        field:
          lexeme: f
          span:
            start: 172
            end: 173
          kind: Ident
    arg:
      Lit:
        Record:
          - ident:
              lexeme: v
              span:
                start: 175
                end: 176
              kind: Ident
            value:
              Use:
                ident:
                  lexeme: v
                  span:
                    start: 177
                    end: 178
                  kind: Ident
- Lit:
    Number: 10
- Lit:
    Bool: true
- Lit:
    Bool: false
- Lit:
    String: "\"test\""
- Lit:
    Record:
      - ident:
          lexeme: v
          span:
            start: 207
            end: 208
          kind: Ident
        value:
          Use:
            ident:
              lexeme: v
              span:
                start: 209
                end: 210
              kind: Ident
- Use:
    ident:
      lexeme: v
      span:
        start: 213
        end: 214
      kind: Ident
- Use:
    ident:
      lexeme: v
      span:
        start: 219
        end: 220
      kind: Ident

